Mathlib/Data/MvPolynomial/Funext.lean

```diff
@@ -0,0 +1,103 @@
+/-
+Copyright (c) 2020 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+
+! This file was ported from Lean 3 source module data.mv_polynomial.funext
+! leanprover-community/mathlib commit da01792ca4894d4f3a98d06b6c50455e5ed25da3
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Polynomial.RingDivision
+import Mathlib.Data.MvPolynomial.Rename
+import Mathlib.RingTheory.Polynomial.Basic
+
+/-!
+## Function extensionality for multivariate polynomials
+
+In this file we show that two multivariate polynomials over an infinite integral domain are equal
+if they are equal upon evaluating them on an arbitrary assignment of the variables.
+
+# Main declaration
+
+* `MvPolynomial.funext`: two polynomials `φ ψ : MvPolynomial σ R`
+  over an infinite integral domain `R` are equal if `eval x φ = eval x ψ` for all `x : σ → R`.
+
+-/
+
+
+namespace MvPolynomial
+
+variable {R : Type _} [CommRing R] [IsDomain R] [Infinite R]
+
+set_option maxHeartbeats 1000000 in
```

Comment:
If this line is absent, there are timeouts. Should I leave it this way?

Reply:
At minimum, needs to have a `-- Porting note`

Reply:
I'm trying to look at this now, but waiting on oleans / `cache get!` problems.

Two possibilities that I'd be okay with:
1) Leave it in, with a porting note
2) Split out two separate lemmas `funext_fin_zero` and `funext_fin_succ`.

Reply:
Okay, the answer here is that this proof should never have been in mathlib3 in the first place. There is a lemma that should have been factored out:

```
theorem eval_eval_finSuccEquiv (f : MvPolynomial (Fin (n + 1)) R) :
    (eval x) (Polynomial.eval q (finSuccEquiv R n f)) = eval (Fin.cons (eval x q) x) f := sorry

private theorem funext_fin {n : ℕ} {p : MvPolynomial (Fin n) R}
    (h : ∀ x : Fin n → R, eval x p = 0) : p = 0 := by
  induction' n with n ih
  · apply (MvPolynomial.isEmptyRingEquiv R (Fin 0)).injective
    rw [RingEquiv.map_zero]
    convert h finZeroElim
  · apply (finSuccEquiv R n).injective
    simp only [AlgEquiv.map_zero]
    refine Polynomial.funext fun q => ?_
    rw [Polynomial.eval_zero]
    apply ih fun x => ?_
    calc _ = _ := eval_eval_finSuccEquiv p
         _ = 0 := h _
```

`eval_eval_finSuccEquiv` needs to be proved, and moved to `Mathlib.Data.MvPolynomial.Equiv`, after `eval_eq_eval_mv_eval'`.

In @jcommelin's defence, when he wrote this proof in https://github.com/leanprover-community/mathlib/pull/4196 the API in `Mathlib.Data.MvPolynomial.Equiv` wasn't as developed.

@Parcly-Taxel, do you want to tackle this? (Up to you of course.)

Reply:
@Parcly-Taxel, I went down the rabbit hole and proved the theorem. There are some `TODO`s about things that need to be moved, still.

Reply:
@semorrison Should we make a mathlib3 PR that moves the theorems to their proper files and rewrites `funext_fin` accordingly?

Reply:
I think that would be a great idea, if you're up for it

Reply:
I've done this now.